[#10] 커서 페이지네이션 구현

mahmunsen opened the pull request:
### 커서 페이지네이션 구현

- jpa 없이 ConcurrentHashMap 상속한 리포지토리로 간략 구현
- 추후 jpa로 구현할 예정

mahmunsen committed:
#2 [chore] @EnableSwagger2 어노테이션 옮김(ApiApplication.java -> SwaggerConfig)

mahmunsen committed:
#1 [chore] god-care-api의 build.gradle 수정
- 스웨거 옯김(god-care-api -> root)
- test 관련 dependencies 옮김(root -> god-care-api)

mahmunsen committed:
#1 [chore] god-care-core, god-care-external 모듈 삭제

mahmunsen committed:
#1 [chore] root 모듈 build.gradle 수정
- 필요하지 않은 dependency 주석 제거
- spring-cloud-dependency 제거
- build 전 설정하는 buildscript 제거

mahmunsen committed:
#3 [feat] 공통 에러 및 응답 코드 작성
- god-care-common에 공통 에러 및 응답 코드 부분에 해당하는 ErrorCode, ApiUtils, Response, CustomException 클래스 등을 작성하였습니다.

mahmunsen committed:
#1 [chore] 각 모듈에서 build.gradle 수정
- root: 호환성 위해서 로그용 slf4j 버전 다운
- god-care-api: external 모듈 제거, 테스트 위한mockito-core dependency 추가
- god-care-core: product 클래스 담을 core 모듈 재생성, god-care-common에 의존성
- settings.gradle: external 모듈 제거

mahmunsen committed:
#3 [feat] ApiUtils 파일 수정
- 안쓰이는 success 메소드 제거
- ResponseEntity 형태로 반환되게끔 수정

mahmunsen committed:
#1 [chore] ApiApplication 클래스 수정
- scanBasePackages에서 external 모듈 디렉토리 제거

mahmunsen committed:
#4 [feat] ProductController에서 상품 등록 api 기능 초안 개발
- ResisterProductRequest(DTO), Product 클래스 필드 2개만 추가하여 간략 개발

mahmunsen committed:
#4 [test] ProductController의 상품 등록 기능에 대한 테스트 코드 작성
- MockMvc 사용, perform으로 http 요청
- repository 사용 대신 HashMap 사용

mahmunsen committed:
#1 [chore] 불필요한 서브 모듈 삭제
- core 모듈 삭제 및 core에 들어있던 service -> api 모듈로 이동
- common 모듈에서 dto 제외한 나머지는 api 모듈로 이동
- settings.gradle, build.gradle 그에 맞게 수정

mahmunsen committed:
#1 [chore] 불필요한 서브 모듈 삭제
- core 모듈 삭제 및 core에 들어있던 service -> api 모듈로 이동
- common 모듈에서 dto 제외한 나머지는 api 모듈로 이동
- settings.gradle, build.gradle 그에 맞게 수정

mahmunsen committed:
Merge remote-tracking branch 'origin/feature/3_product' into feature/3_product

# Conflicts:
#	build.gradle
#	god-care-api/build.gradle
#	god-care-api/src/main/java/com/godcare/api/ApiApplication.java
#	god-care-common/build.gradle
#	settings.gradle

mahmunsen committed:
#0 [docs] SwaggerConfig documentationType 수정

mahmunsen committed:
#0 [chore] ApiApplication EntityScan import 제거

mahmunsen committed:
#4 [test] Product 컨트롤러 테스트 수정
- jsonPath -> response 객체로 받아서 검증

mahmunsen committed:
#4 [feat] ProductService 작성
- 상품등록, 조회, 전체조회, 수정 및 삭제

mahmunsen committed:
#4 [feat] ProductRepository 작성
- 상품등록, 조회, 전체조회, 수정 및 삭제

mahmunsen committed:
#4 [feat] Product(엔티티) 수정

mahmunsen committed:
#4 [feat] product 관련 dto 생성
- common모듈에 생성
- 필드는 일단 간략하게 구성

mahmunsen committed:
#4 [feat] ProductController 상품 관련 api 추가 생성

mahmunsen committed:
#0 [docs] SwaggerConfig 수정
- documentationType 변경

mahmunsen committed:
#4 [feat] Product 관련 파일 수정

mahmunsen committed:
#4 [test] service 테스트 작성

mahmunsen committed:
#4 [test] ProductServiceTest 업로드
- @Mock, @InjectMocks 이용한 서비스 테스트 코드

mahmunsen committed:
#3 [chore] ApiUtils 제거

mahmunsen committed:
#1 [chore] build.gradle 수정
- dependency, plugin 등: 알파벳 순으로 정렬, single quotes 사용

mahmunsen committed:
#3 [refactor] Response 클래스 내용 수정
- success, fail로 구분

mahmunsen committed:
#3 [refactor] ErrorCode의 getCode() 수정
- int -> HttpStatus

mahmunsen committed:
#3 [refactor] CustomException 수정

mahmunsen committed:
#3 [feat] CustomExceptionAdvice 생성
- global exception handler 생성

mahmunsen committed:
#3 [refactor] ProductErrorCode 수정
- int -> HttpStatus

mahmunsen committed:
#4 [refactor] ProductServiceTest 수정
- @BeforeEach 코드블럭 삭제

mahmunsen committed:
#4 [refactor] ProductApiTest 수정
- response 객체로 변환

mahmunsen committed:
#3 [refactor] Response 수정
- 컨트롤러 테스트코드에서 역직렬화가 필요하여 기본 생성자 추가

mahmunsen committed:
#4 [feat] ProductApiTest2 생성
- 컨트롤러 테스트 코드, 조회 작성

mahmunsen committed:
#4 [refactor] ProductController 수정
- ApiUtils 제거
- @ResponseStatus 사용

mahmunsen committed:
#4 [refactor] ProductRepository 수정
- 불필요한 import 제거

mahmunsen committed:
#4 [feat] ResisterProductResponse 응답 dto 생성

mahmunsen committed:
#4 [feat] 상품 리스트 조회 테스트 생성

mahmunsen committed:
#4 [feat] 상품 수정 테스트 생성

mahmunsen committed:
#4 [refactor] ProductController 상품 수정 api 수정

mahmunsen committed:
#4 [feat] 상품 삭제 테스트 생성

mahmunsen committed:
#3 [refactor] CustomException, ErrorCode, ProductErrorCode 파일 삭제
- CustomException 방식 변경으로 인한 삭제

mahmunsen committed:
#3 [refactor] CustomExceptionAdvice 수정 맟 ProductNotFoundException 생성
- runtimeException extends한 customException 클래스 생성

mahmunsen committed:
#3 [refactor] Response 수정
- message 필드 삭제(fail시 메시지는 data로 포함)

mahmunsen committed:
#4 [refactor] requestDTO, responseDTO 수정
- 불필요한 어노테이션 제거
- @NonNull 추가

mahmunsen committed:
#4 [refactor] Product, ProductService 수정
- builder 패턴 제거

mahmunsen committed:
#4 [refactor] ProductRepository 수정
- ConcurrentHashMap으로 대체
- Atomic 키워드로 수정

mahmunsen committed:
#4 [refactor] ProductController 응답 부분 수정

mahmunsen committed:
#5 [chore] jpa 사용 위한 설정

mahmunsen committed:
#5 [chore] jpa 사용 위한 설정

mahmunsen committed:
#5 [chore] jpa 쓰기 위한 dependency 추가 및 jasypt 추가

mahmunsen committed:
#5 [chore] jpa 쓰기 위한 dependency 추가 및 jasypt 추가

mahmunsen committed:
#5 [docs] JasyptConfig 작성
- secret_key_source 암호화 위한 configuration 생성

mahmunsen committed:
#5 [docs] JasyptConfig 작성
- secret_key_source 암호화 위한 configuration 생성

mahmunsen committed:
#5 [feat] product, category 엔티티 생성

mahmunsen committed:
#5 [feat] product, category 엔티티 생성

mahmunsen committed:
#5 [feat] PageResponse 생성
- 페이징 응답 위한 클래스

mahmunsen committed:
#5 [feat] PageResponse 생성
- 페이징 응답 위한 클래스

mahmunsen committed:
#5 [feat] 커서 페이지네이션(jpa없이) 간략 구현

mahmunsen committed:
#5 [feat] 커서 페이지네이션(jpa없이) 간략 구현

mahmunsen committed:
#5 [feat] jpa 사용 위한 리포지토리 생성

mahmunsen committed:
#5 [feat] jpa 사용 위한 리포지토리 생성

f-lab-lyan commented on god-care-api/src/main/java/com/godcare/api/repository/ProductHashMapRepository.java:
```diff
@@ -0,0 +1,86 @@
+    public List<Product> findAllByOrderByIdDesc(Pageable pageable) {
+        Result result = getResult(pageable);
+
+        return entrySet().stream()
+                .skip(result.start)
+                .limit(result.end - result.start)
+                .sorted(Map.Entry.<Long, Product>comparingByKey().reversed())
+                .map(Map.Entry::getValue)
+                .collect(Collectors.toList());
```
```suggestion
        return keySet().stream()
                .sorted(Key.reverseOrder())
                .skip(result.start)
                .limit(result.end - result.start)
                .map(k -> getNode(k))
                .collect(Collectors.toList());
```

mahmunsen committed:
Merge pull request #8 from mahmunsen/feature/3_product

[#7] Product api 개발

mahmunsen committed:
Merge remote-tracking branch 'origin/feature/4_cursorPaging' into feature/4_cursorPaging

f-lab-lyan commented:
@mahmunsen 민승님, 이미 반영된 변경사항들은 이 PR에서 빼주실 수 있을까요? 정리하고 진행해보죠. :-)

f-lab-lyan commented:
그리고 DB를 올려줘야 할 것 같은데요. 저는 보통 docker-compose를 추천하는데요. 혹시 다른 더 좋은 아이디어가 있으실까요?

mahmunsen committed:
#5 [refactor] jpa 없는 커서 페이지네이션 간략 구현 수정
- 추후 jpa 이용한 커서 페이지네이션 부분과 구분하기 위해 아래 클래스 이름 변경
- 기존 엔티티 이름 Product -> ProductV1
- 기존 서비스 이름 ProductService -> ProductV1Service 로 변경 등

f-lab-lyan commented on god-care-api/src/main/java/com/godcare/api/repository/ProductHashMapRepository.java:
```diff
@@ -1,86 +1,67 @@
-                .collect(Collectors.toList());
+    public List<ProductV1> findAllByOrderByIdDesc(Pageable pageable) {
+        List<ProductV1> productList = new ArrayList<>();
+        for(Long key: storage.keySet()){
```
https://www.baeldung.com/java-treemap 처음에 저장할 때, 거꾸로 저장하면, 조회할 때, 다시 거꾸로 정렬할 필요가 없겠죠.

f-lab-lyan commented on god-care-api/src/main/java/com/godcare/api/repository/ProductHashMapRepository.java:
```diff
@@ -0,0 +1,67 @@
+        return productList.subList(end - pageable.getPageSize(), end);
+    }
+    public List<ProductV1> findByIdLessThanOrderByIdDesc(Long id, Pageable pageable) {
+        List<ProductV1> productList = new ArrayList<>();
+        for(Long key: storage.keySet()){
+            if(key>id){
+                productList.add(storage.get(key));
+            }
+        }
+        Collections.reverse(productList);
+        int end = productList.size();
+        return productList.subList(end - pageable.getPageSize(), end);
```
```suggestion
        // id = 50, storage.keySet().size = 1000, pageSize = 10
        List<ProductV1> productList = new ArrayList<>();
        for(Long key: storage.keySet()){
            if(key>id){ // 51 ... 1000
                productList.add(storage.get(key)); // 51...1000
            }
        }
        Collections.reverse(productList); // 1000...51
        int end = productList.size(); // 950
        // end - pageable.getPageSize() = 950 - 10 = 940
        return productList.subList(end - pageable.getPageSize(), end); // subList(940, 950)
        // 60...51
```

```suggestion
        // id = 50, storage.keySet().size = 1000, pageSize = 10
        List<ProductV1> productList = new ArrayList<>();
        for(Long key: storage.keySet()){
            if(key>id){ // 51 ... 1000
                productList.add(storage.get(key)); // 51...1000
                if (productList.size() > pageable.getPageSize()) {
                  // Collections.sort(productList, Comparing.reverse());
                  return productList;
                }
            }
        }
```

mahmunsen committed:
#5 [refactor] jpa 없는 간략 커서 페이징 수정

f-lab-lyan commented on god-care-api/src/main/java/com/godcare/api/repository/ProductHashMapRepository.java:
```diff
@@ -1,62 +1,61 @@
+    public List<ProductV0> findByIdGreaterThanOrderByIdDesc(Long id, Pageable pageable) {
+        List<ProductV0> productList = new ArrayList<>();
+        for (Long key : storage.keySet()) {
+            if (id < key && key <= id + pageable.getPageSize()) {
```
조금 더 성능을 올려보면, productList에 원하는 만큼 데이터가 쌓이면, for를 끝까지 돌 필요는 없겠죠? :-) 이건 다른 PR에서 작업해보죠.
